aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -8,12 +8,84 @@
 #include <ATen/SparseCsrTensorImpl.h>
 #include <ATen/SparseCsrTensorUtils.h>
 #include <ATen/SparseTensorImpl.h>
+#include <ATen/InitialTensorOptions.h>
 
 namespace at {
 namespace native {
 
 using namespace at::sparse_csr;
 
+namespace {
+
+
+} // end anonymous namespace
+
+void _validate_sparse_csr_tensor_args(const Tensor& crow_indices, const Tensor& col_indices, const Tensor& values, IntArrayRef size) {
+  // Layout Invariants
+  TORCH_CHECK(
+      col_indices.layout() == kStrided && col_indices.is_contiguous(),
+      "expected col_indices to be a strided and contiguous tensor");
+
+  TORCH_CHECK(
+      crow_indices.layout() == kStrided && crow_indices.is_contiguous(),
+      "expected crow_indices to be a strided and contiguous tensor");
+
+  TORCH_CHECK(
+      values.layout() == kStrided && values.is_contiguous(),
+      "expected values to be a strided and contiguous tensor");
+
+  // Shape and Strides invariants
+  TORCH_CHECK(
+      size.size() == 2,
+      "size of a CSR tensor must be of length 2, but got: ",
+      size.size());
+  TORCH_CHECK(
+      crow_indices.dim() == 1,
+      "crow_indices must have dim=1 but got crow_indices.dim()=",
+      crow_indices.dim());
+  TORCH_CHECK(
+      col_indices.dim() == 1,
+      "col_indices must have dim=1 but got col_indices.dim()=",
+      col_indices.dim());
+  TORCH_CHECK(
+      values.dim() == 1,
+      "values must have dim=1 but got values.dim()=",
+      values.dim());
+  // Note, this check also enforces `crow_indices.numel() >= 1`
+  TORCH_CHECK(
+      crow_indices.numel() == (size[0] + 1),
+      "crow_indices.numel() must be size(0) + 1, but got: ",
+      crow_indices.numel());
+  TORCH_CHECK(
+      col_indices.size(0) == values.size(0),
```

Comment:
Nit: If `t` is `crow_indices`, `col_indices`, or `values`, then we have `t.size(0) == t.numel()`. A quick benchmark shows that `t.numel()` is about 3-4x faster than `t.size(0)`. Although the timings are in the order of microseconds, I suggest using `numel()` instead of `size(0)` for consistency here and below.